File: src/program/hybrid_access/base/ring_queue.lua
---@diagnostic disable: inject-field
module(..., package.seeall)

local ffi = require("ffi")

require("core.packet_h")
require("program.hybrid_access.base.ring_queue_h")

local delayer_buffer_t = ffi.typeof("struct delay_buffer")

local BUFFER_LENGTH = ffi.C.QUEUE_SIZE

RingQueue = {}

function RingQueue:new()
    local o = {}
    o.buffer = ffi.new(delayer_buffer_t)
    o.default_sending_time = ffi.new("uint64_t", -1)

    setmetatable(o, self)
    self.__index = self
    return o
end

function RingQueue:dequeue()
    -- assert(not self:empty())
    local buffer = self.buffer
    local timed_pkt = buffer.packets[buffer.read]
    
    local pkt_pointer = timed_pkt.packet
    
    timed_pkt.packet = nil
    buffer.read = (buffer.read + 1) % BUFFER_LENGTH

    return pkt_pointer
end

function RingQueue:enqueue(pkt, sending_time)
    -- assert(not self:full())
    local buffer = self.buffer
    local timed_pkt = buffer.packets[buffer.write]
    timed_pkt.packet = pkt
    timed_pkt.sending_time = sending_time

    buffer.write = (buffer.write + 1) % BUFFER_LENGTH
end

function RingQueue:is_empty()
    return self.buffer.read == self.buffer.write
end

function RingQueue:is_full()
    return self.buffer.read == (self.buffer.write + 1) % BUFFER_LENGTH
end

function RingQueue:nreadable()
    local buffer = self.buffer
    if buffer.read > buffer.write then
       return buffer.write + BUFFER_LENGTH - buffer.read
    else
       return buffer.write - buffer.read
    end
end

function RingQueue:peek_time()
    if self:is_empty() then
        return self.default_sending_time
    else
        return self.buffer.packets[self.buffer.read].sending_time
    end
end

File: src/program/hybrid_access/base/ring_queue.c
#include <stdint.h>
#include <stdlib.h>
#include "ring_queue.h"

// Initialize the delay buffer
struct ring_buffer* create_buffer() {
    struct ring_buffer *buffer = (struct ring_buffer*) malloc(sizeof(struct ring_buffer));
    if (buffer) {
        buffer->read = 0;
        buffer->write = 0;
    }
    return buffer;
}

// Check if the buffer is empty
inline int is_empty(struct ring_buffer *buffer) {
    return buffer->read == buffer->write;
}

// Check if the buffer is full
inline int is_full(struct ring_buffer *buffer) {
    return ((buffer->write + 1) & (QUEUE_SIZE - 1)) == buffer->read;
}

// Enqueue a timed packet to the buffer
// not full check needs to be done
void enqueue(struct ring_buffer *buffer, struct packet *pkt, uint64_t sending_time) {
    // assert not full
    buffer->packets[buffer->write].packet = pkt;
    buffer->packets[buffer->write].sending_time = sending_time;

    buffer->write = (buffer->write + 1) & (QUEUE_SIZE - 1);
}

// Dequeue a timed packet from the buffer
// not empty check needs to be done
struct timed_packet* dequeue(struct ring_buffer *buffer) {
    // assert not empty
    struct timed_packet *pkt = &(buffer->packets[buffer->read]);
    buffer->read = (buffer->read + 1) & (QUEUE_SIZE - 1);
    return pkt;
}

// Peek at the sending time of the next timed packet to be read without removing it from the buffer
uint64_t peek_time(struct ring_buffer *buffer) {
    if (is_empty(buffer)) {
        // Buffer is empty
        return UINT64_MAX;
    }
    return buffer->packets[buffer->read].sending_time;
}

// Amount of readable packets
int buffer_size(struct ring_buffer *buffer) {
    if (buffer->write > buffer->read) {
        return buffer->write - buffer->read;
    }
    return buffer->write + QUEUE_SIZE - buffer->read;
}

// Destroy the delay buffer and free its memory
void free_buffer(struct ring_buffer *buffer) {
    if (buffer) {
        free(buffer);
    }
}